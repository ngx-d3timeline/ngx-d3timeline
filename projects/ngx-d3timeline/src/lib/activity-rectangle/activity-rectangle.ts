import { identifier } from '../core/types';

export interface ActivityRectangle {
  id: identifier;
  title: string;
  showTitle: boolean;
  transform: string;
  width: number;
  height: number;
  fontSize: number;
}
