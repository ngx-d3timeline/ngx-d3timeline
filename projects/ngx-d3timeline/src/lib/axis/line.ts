import { Orientation } from '../orientation';

export interface Line {
  x1: number;
  x2: number;
  y1: number;
  y2: number;
}

export function createLine(
  x1: number,
  y1: number,
  length: number,
  orientation: Orientation
) {
  return orientation === Orientation.Vertical
    ? createVerticalLine(x1, y1, length)
    : createHorizontalLine(x1, y1, length);
}

export function createVerticalLine(x1: number, y1: number, length: number) {
  return {
    x1,
    y1,
    x2: x1,
    y2: y1 + length
  };
}

export function createHorizontalLine(x1: number, y1: number, length: number) {
  return {
    x1,
    y1,
    x2: x1 + length,
    y2: y1
  };
}
