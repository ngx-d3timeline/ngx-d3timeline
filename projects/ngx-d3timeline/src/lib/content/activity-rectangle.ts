import { identifier } from '../types';

export interface ActivityRectangle {
  id: identifier;
  title: string;
  showTitle: boolean;
  transform: string;
  width: number;
  height: number;
  fontSize: number;
}
