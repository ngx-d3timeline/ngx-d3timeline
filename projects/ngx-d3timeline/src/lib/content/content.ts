export interface EventRectangle {
  id: string | number;
  title: string;
  transform: string;
  width: number;
  height: number;
}
