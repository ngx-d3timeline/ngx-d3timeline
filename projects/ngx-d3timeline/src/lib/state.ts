import { TimelineEvent } from './timeline-event';
import { TimelineView } from './view/timeline-view';
import { AxisOrientations } from './axis-orientations';

export class State {
  data: TimelineEvent[];
  orientations: AxisOrientations;
  view: TimelineView;
}
