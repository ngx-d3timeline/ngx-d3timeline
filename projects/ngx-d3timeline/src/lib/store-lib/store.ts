import { Injectable } from '@angular/core';
import { ReplaySubject, Observable } from 'rxjs';
import { shareReplay, scan, map } from 'rxjs/operators';
import { Action } from './action';
import { reducer } from '../store/reducer';
import { Selector } from './selector/selector';

@Injectable({ providedIn: 'root' })
export class Store {
  private readonly replayBufferSize = 100;
  private actionsSubject = new ReplaySubject<Action>(this.replayBufferSize);

  state$ = this.actionsSubject.pipe(scan(reducer, undefined), shareReplay());

  dispatch(action: Action): void {
    this.actionsSubject.next(action);
  }

  select<TResult>(selector: Selector<TResult>): Observable<TResult> {
    return this.state$.pipe(map(state => selector.execute(state)));
  }
}
