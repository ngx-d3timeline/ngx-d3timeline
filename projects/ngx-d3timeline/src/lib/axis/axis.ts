import { TickMark } from '../tick-mark/tick-mark';
import { Line } from './line';

export interface Axis {
  tickMarks: TickMark[];
  line: Line;
}
