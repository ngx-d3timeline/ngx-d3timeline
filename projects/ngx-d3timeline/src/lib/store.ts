import { Injectable } from '@angular/core';
import { BehaviorSubject, combineLatest, Observable } from 'rxjs';
import { TimelineEvent } from './timeline-event';
import { TimelineView } from './view/timeline-view';
import { Orientation } from './orientation';
import { State } from './state';
import { map } from 'rxjs/operators';

@Injectable({ providedIn: 'root' })
export class Store {
  private readonly margin = 50;

  private dataSubject = new BehaviorSubject<TimelineEvent[]>(null);
  private viewSubject = new BehaviorSubject<TimelineView>(null);
  private orientationSubject = new BehaviorSubject<Orientation>(
    Orientation.Vertical
  );

  state$: Observable<State> = combineLatest([
    this.dataSubject.asObservable(),
    this.viewSubject.asObservable(),
    this.orientationSubject.asObservable()
  ]).pipe(
    map(([data, view, timelineOrientation]) => ({
      data,
      view,
      timelineOrientation
    }))
  );

  setData(data: TimelineEvent[]) {
    this.dataSubject.next(data);
  }

  setView([width, height]: [number, number]) {
    this.viewSubject.next(
      new TimelineView({ width, height, margin: this.margin })
    );
  }

  setOrientation(orientation: Orientation) {
    this.orientationSubject.next(orientation);
  }
}
