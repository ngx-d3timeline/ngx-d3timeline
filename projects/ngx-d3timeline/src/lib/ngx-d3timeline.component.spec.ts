import { async, ComponentFixture, TestBed } from '@angular/core/testing';

import { NgxD3timelineComponent } from './ngx-d3timeline.component';
import { ContentComponent } from './content/content.component';
import { ActivityRectangleComponent } from './content/activity-rectangle.component';
import { Store } from './store-lib/store';
import { initialState } from './store/state';
import { of } from 'rxjs';
import { LineComponent } from './line.component';
import { AxisTickMarkComponent } from './axis/axis-tick-mark.component';
import { ClipPathComponent } from './content/clip-path.component';
import { Component, Input } from '@angular/core';
import { Axis } from './axis/axis';
import { NgxD3TimelineService } from './ngx-d3timeline.service';
import { View } from './view/view';

@Component({
  selector: '[ngx-d3timeline-axis]',
  template: `
    <svg:g class="axis-group"></svg:g>
  `
})
class FakeAxisComponent {
  @Input() axis: Axis;
}

@Component({
  selector: '[ngx-d3timeline-content]',
  template: `
    <svg:g class="content"></svg:g>
  `
})
class FakeContentComponent {}

describe('NgxD3timelineComponent', () => {
  let component: NgxD3timelineComponent;
  let fixture: ComponentFixture<NgxD3timelineComponent>;
  let timeline: NgxD3TimelineService;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [
        NgxD3timelineComponent,
        FakeContentComponent,
        FakeAxisComponent
      ],
      providers: [
        {
          provide: NgxD3TimelineService,
          useValue: {
            timeAxis$: jest.fn(),
            resourceAxis$: jest.fn(),
            content$: jest.fn(),
            lastDraggedActivity$: jest.fn(),
            setupZoom: jest.fn()
          }
        }
      ]
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(NgxD3timelineComponent);
    timeline = TestBed.inject(NgxD3TimelineService);
    timeline.lastDraggedActivity$ = of(null);

    component = fixture.componentInstance;
  });

  it('should not render if view null', () => {
    timeline.view$ = of(null);

    fixture.detectChanges();

    expect(fixture.nativeElement).toMatchSnapshot();
  });

  it('should render correctly', () => {
    timeline.view$ = of(new View([800, 600]));
    timeline.resourceAxis$ = of(null);
    timeline.timeAxis$ = of(null);

    fixture.detectChanges();

    expect(fixture.nativeElement).toMatchSnapshot();
  });
});
