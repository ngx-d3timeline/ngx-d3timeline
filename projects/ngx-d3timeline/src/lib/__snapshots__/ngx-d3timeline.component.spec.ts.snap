// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NgxD3timelineComponent should not render if view null 1`] = `
<div
  id="root0"
>
  
</div>
`;

exports[`NgxD3timelineComponent should render correctly 1`] = `
<div
  id="root1"
>
  
  <svg
    class="ngx-d3timeline"
    height="600"
    width="800"
  >
    <g
      class="resources-axis"
      ngx-d3timeline-axis=""
    >
      <g
        class="axis-group"
      />
    </g>
    <g
      class="time-axis"
      ngx-d3timeline-axis=""
    >
      <g
        class="axis-group"
      />
    </g>
    <g
      ngx-d3timeline-content=""
    >
      <g
        class="content"
      />
    </g>
  </svg>
</div>
`;
